### PR TITLE
feat(core): remove @nrwl/workspace:lib

### PR DESCRIPTION
**Pull request by meeroslav**
## Related Issue(s)

Fixes #

### CODE REVIEW

**Review by barbados-clemens — APPROVED** [review]

**Review by nartc — APPROVED** [review]
the day has come!

**Review by AgentEnder — APPROVED** [review]

File: e2e/nx-misc/src/workspace.test.ts [hunk 1]
```diff
@@ -27,12 +27,12 @@ describe('Workspace Tests', () => {
 
   afterAll(() => cleanupProject());
 
-  describe('@nrwl/workspace:library', () => {
+  describe('@nrwl/js:library', () => {
```
Comment (FrozenPandaz) [inline e2e/nx-misc/src/workspace.test.ts:30]: These should probably be moved to `e2e-js`

File: packages/nx/src/command-line/generate.ts [hunk 2]
```diff
@@ -235,7 +235,7 @@ async function convertToGenerateOptions(
   }
 
   if (!collectionName) {
-    throwInvalidInvocation(['@nrwl/workspace:library']);
+    throwInvalidInvocation(['@nrwl/js:library']);
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/generate.ts:238]: Is this working as expected?

File: packages/workspace/src/generators/move/lib/update-cypress-config.spec.ts [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
   Tree,
   writeJson,
 } from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { updateCypressConfig } from './update-cypress-config';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/update-cypress-config.spec.ts:15]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/move/lib/update-imports.spec.ts [hunk 4]
```diff
@@ -1,10 +1,18 @@
...
+  Tree,
+} from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
 import { Schema } from '../schema';
 import { updateImports } from './update-imports';
 import { normalizeSchema } from './normalize-schema';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/update-imports.spec.ts:14]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/move/lib/update-jest-config.spec.ts [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
+  readProjectConfiguration,
+  Tree,
+} from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { updateJestConfig } from './update-jest-config';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/update-jest-config.spec.ts:12]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/move/lib/update-package-json.spec.ts [hunk 6]
```diff
@@ -1,9 +1,17 @@
...
+  Tree,
+  writeJson,
+} from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { updatePackageJson } from './update-package-json';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/update-package-json.spec.ts:13]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/move/lib/update-project-root-files.spec.ts [hunk 7]
```diff
@@ -1,9 +1,16 @@
...
+  readProjectConfiguration,
+  Tree,
+} from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { updateProjectRootFiles } from './update-project-root-files';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/update-project-root-files.spec.ts:12]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/move/lib/update-readme.spec.ts [hunk 8]
```diff
@@ -1,10 +1,12 @@
-import { Tree } from '@nrwl/devkit';
+import { ensurePackage, NX_VERSION, Tree } from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
 import { join } from 'path';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { updateReadme } from './update-readme';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/update-readme.spec.ts:8]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/move/lib/update-storybook-config.spec.ts [hunk 9]
```diff
@@ -1,9 +1,16 @@
...
+  readProjectConfiguration,
+  Tree,
+} from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { updateStorybookConfig } from './update-storybook-config';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/update-storybook-config.spec.ts:12]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/move/move.spec.ts [hunk 10]
```diff
@@ -1,7 +1,8 @@
-import { readJson, Tree } from '@nrwl/devkit';
+import { ensurePackage, NX_VERSION, readJson, Tree } from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
 import { moveGenerator } from './move';
-import { libraryGenerator } from '../library/library';
+
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/move.spec.ts:5]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/remove/lib/check-dependencies.spec.ts [hunk 11]
```diff
@@ -8,7 +8,7 @@ import {
 import { Schema } from '../schema';
 import { checkDependencies } from './check-dependencies';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
+// import { libraryGenerator } from '../../library/library';
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/remove/lib/check-dependencies.spec.ts:11]: Remove?

File: packages/workspace/src/generators/remove/lib/remove-project.spec.ts [hunk 12]
```diff
@@ -1,9 +1,16 @@
...
+  readProjectConfiguration,
+  Tree,
+} from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
 import { Schema } from '../schema';
-import { libraryGenerator } from '../../library/library';
 import { removeProject } from '@nrwl/workspace/src/generators/remove/lib/remove-project';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/remove/lib/remove-project.spec.ts:12]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/remove/lib/update-jest-config.spec.ts [hunk 13]
```diff
@@ -1,12 +1,19 @@
...
 
 import { readFileSync } from 'fs';
 import { join } from 'path';
 
 import { Schema } from '../schema';
 import { updateJestConfig } from './update-jest-config';
-import { libraryGenerator } from '../../library/library';
+
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/remove/lib/update-jest-config.spec.ts:16]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

File: packages/workspace/src/generators/remove/lib/update-tsconfig.spec.ts [hunk 14]
```diff
@@ -7,7 +9,9 @@ import {
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
 import { Schema } from '../schema';
 import { updateTsconfig } from './update-tsconfig';
-import { libraryGenerator } from '../../library/library';
+
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/remove/lib/update-tsconfig.spec.ts:14]: You should be able to use 
```ts
// nx-ignore-next-line
const { libraryGenerator } = require('@nrwl/js');
```

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by xiongemi — APPROVED** [review]

File: packages/nx/src/command-line/generate.ts (hunk 2 above)
Comment (meeroslav) [reply packages/nx/src/command-line/generate.ts]: Actually not, the only way we might not have a `collectionName` is by running a `workspace-generator`, which at the moment fails anyways with:
```
 >  NX   Could not find any generators named "my-generator"
```

I will remove that block

cc @AgentEnder should this still be possible or should the `workspace-generator` generator be removed in v16?:
1. npx nx generate @nrwl/workspace:workspace-generator my-generator
2. npx nx generate my-generator --dry-run
3. Fails with the above error

**Review by mandarini — APPROVED** [review]

File: packages/workspace/src/generators/move/lib/move-project-files.spec.ts [hunk 15]
```diff
@@ -1,20 +1,24 @@
...
   readProjectConfiguration,
   Tree,
 } from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { moveProjectFiles } from './move-project-files';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
[the remark this answers is not in the document]
Comment (meeroslav) [reply packages/workspace/src/generators/move/lib/move-project-files.spec.ts:13]: Actually not; using `require` also produces circular-dependency on build:
Screenshot 2023-04-06 at 12 32 50

File: packages/workspace/src/generators/move/lib/update-cypress-config.spec.ts (hunk 3 above)
Comment (meeroslav) [reply packages/workspace/src/generators/move/lib/update-cypress-config.spec.ts:15]: See my comment above

File: packages/workspace/src/generators/move/lib/update-eslintrc-json.spec.ts [hunk 16]
```diff
@@ -1,15 +1,19 @@
...
   updateJson,
 } from '@nrwl/devkit';
 import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
 import { Linter } from '../../../utils/lint';
-import { libraryGenerator } from '../../library/library';
 import { NormalizedSchema } from '../schema';
 import { updateEslintrcJson } from './update-eslintrc-json';
 
+// avoid circular deps
+const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
[the remark this answers is not in the document]
Comment (meeroslav) [reply packages/workspace/src/generators/move/lib/update-eslintrc-json.spec.ts:15]: See my comment above

File: packages/workspace/src/generators/remove/lib/remove-project.spec.ts (hunk 12 above)
Comment (meeroslav) [reply packages/workspace/src/generators/remove/lib/remove-project.spec.ts:12]: See my comment above

File: packages/workspace/src/generators/remove/lib/update-jest-config.spec.ts (hunk 13 above)
Comment (meeroslav) [reply packages/workspace/src/generators/remove/lib/update-jest-config.spec.ts:16]: See my comment above

File: packages/workspace/src/generators/remove/lib/update-tsconfig.spec.ts (hunk 14 above)
Comment (meeroslav) [reply packages/workspace/src/generators/remove/lib/update-tsconfig.spec.ts:14]: See my comment above

File: packages/workspace/src/generators/run-commands/run-commands.spec.ts [hunk 17]
```diff
@@ -14,6 +14,8 @@ describe('run-commands', () => {
       outputs: 'dist/a, dist/b, dist/c',
     };
 
+    // avoid circular deps
+    const { libraryGenerator } = ensurePackage('@nrwl/js', NX_VERSION);
```
[the remark this answers is not in the document]
Comment (meeroslav) [reply packages/workspace/src/generators/run-commands/run-commands.spec.ts:18]: See my comment above

File: e2e/nx-misc/src/workspace.test.ts (hunk 1 above)
Comment (meeroslav) [reply e2e/nx-misc/src/workspace.test.ts]: I moved and combined them with existing tests in the js project

**Review by isaacplmann — APPROVED** [review]

**Review by FrozenPandaz — APPROVED** [review]

**Pull request merged**